src/main/src/main/java/org/geoserver/security/impl/DefaultResourceAccessManager.java

```diff
@@ -705,9 +705,11 @@ private Filter buildInFunctionResourceFilter(
                 }
                 boolean layerAccess = canAccess(user, layerNode);
                 if (layerAccess != wsAccess) {
-                    if (clazz.isAssignableFrom(published.getClass()))
+                    if (ResourceInfo.class.isAssignableFrom(clazz)) {
+                        layerExceptionIds.add(((LayerInfo) published).getResource().getId());
+                    } else {
                         layerExceptionIds.add(published.getId());
-                    else layerExceptionIds.add(((LayerInfo) published).getResource().getId());
+                    }
```

Comment:
The change works (tested manually) and yet I'm perplexed... if I had a situation where I'm listing ResourceInfo.class, and there is a workspace specific LayerGroup that has access different than the workspace, it would still throw a ClassCastException.

Yes, manually reproduced that case. I had to set up the limited role access on a workspace specific layer group, and then asked for the list of all collections in OGC API - Features, and here we go:

```
	at freemarker.ext.beans.BeansWrapper.invokeMethod(BeansWrapper.java:1552)
	at freemarker.ext.beans.BeanModel.invokeThroughDescriptor(BeanModel.java:233)
	at freemarker.ext.beans.BeanModel.get(BeanModel.java:152)
	... 130 more
Caused by: java.lang.ClassCastException: class jdk.proxy2.$Proxy53 cannot be cast to class org.geoserver.catalog.LayerInfo (jdk.proxy2.$Proxy53 is in module jdk.proxy2 of loader 'app'; org.geoserver.catalog.LayerInfo is in unnamed module of loader 'app')
	at org.geoserver.security.impl.DefaultResourceAccessManager.buildInFunctionResourceFilter(DefaultResourceAccessManager.java:709)
	at org.geoserver.security.impl.DefaultResourceAccessManager.getSecurityFilter(DefaultResourceAccessManager.java:562)
	at org.geoserver.security.CatalogFilterAccessManager.getSecurityFilter(CatalogFilterAccessManager.java:178)
	at org.geoserver.security.SecureCatalogImpl.securityFilter(SecureCatalogImpl.java:1721)
	at org.geoserver.security.SecureCatalogImpl.list(SecureCatalogImpl.java:1627)
	at org.geoserver.catalog.impl.AbstractFilteredCatalog.list(AbstractFilteredCatalog.java:933)
	at org.geoserver.catalog.impl.LocalWorkspaceCatalog.list(LocalWorkspaceCatalog.java:588)
	at org.geoserver.catalog.impl.LocalWorkspaceCatalog.list(LocalWorkspaceCatalog.java:569)
	at org.geoserver.ogcapi.v1.features.CollectionsDocument.getCollections(CollectionsDocument.java:55)
```

Reply:
fixed